django/db/models/query.py

```diff
@@ -1581,8 +1582,19 @@ def prefetch_one_level(instances, prefetcher, lookup, level):
 
         if single:
             val = vals[0] if vals else None
-            to_attr = to_attr if as_attr else cache_name
-            setattr(obj, to_attr, val)
+
+            if as_attr:
+                # A to_attr has been given for the prefetch
+                setattr(obj, to_attr, val)
+            elif is_descriptor:
+                # cache_name points to a field name in obj
+                # this field is a descriptor for a related object.
+                setattr(obj, cache_name, val)
+            else:
+                # No to_attr has been given for this prefetch operation
+                # and the cache_name does not point to a descriptor.
+                # store the value of the field directly in the object's field cache.
+                obj._state.cache[cache_name] = val
```

Comment:
Poking into the fields cache directly (without going through the usual field-cache API) feels really ugly, as does this whole `is_descriptor` thing. The code here is just much harder to understand than the previous version (unlike in the rest of this PR).

What if we pushed this responsibility back onto the implementation of `get_prefetch_queryset`? Instead of having the last two items of the returned tuple be `cache_name, is_descriptor`, have it be `set_func`, which should be a callable with the signature `(obj, val)`?

Or if that doesn't work out nicely, can't we at least get the field object here and use the normal field-cache-setting API?